server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -0,0 +1,521 @@
+   */
+  public SegmentIdentifier add(
+      final InputRow row,
+      final Supplier<Committer> committerSupplier
+  ) throws IOException
+  {
+    final SegmentIdentifier identifier = getSegment(row.getTimestamp());
+
+    if (identifier != null) {
+      try {
+        final int numRows = appenderator.add(identifier, row, wrapCommitterSupplier(committerSupplier));
+        if (numRows >= maxRowsPerSegment) {
+          moveSegmentOut(ImmutableList.of(identifier));
+        }
+      }
+      catch (SegmentNotWritableException e) {
+        throw new ISE(e, "WTF?! Segment[%s] not writable when it should have been.", identifier);
+      }
+    }
+
+    return identifier;
+  }
+
+  public int getActiveSegmentCount()
+  {
+    synchronized (activeSegments) {
+      return activeSegments.size();
+    }
+  }
+
+  public List<SegmentIdentifier> getActiveSegments()
+  {
+    synchronized (activeSegments) {
+      return ImmutableList.copyOf(activeSegments.values());
+    }
+  }
+
+  /**
+   * Persist all data indexed through this driver so far. Blocks until complete.
+   *
+   * Should be called after all data has been added through {@link #add(InputRow, Supplier)}.
+   *
+   * @param committer committer representing all data that has been added so far
+   *
+   * @return commitMetadata persisted
+   */
+  public Object persist(final Committer committer) throws InterruptedException
+  {
+    try {
+      log.info("Persisting data.");
+      final long start = System.currentTimeMillis();
+      final Object commitMetadata = appenderator.persistAll(wrapCommitter(committer)).get();
+      log.info("Persisted pending data in %,dms.", System.currentTimeMillis() - start);
+      return commitMetadata;
+    }
+    catch (InterruptedException e) {
+      throw e;
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Publish all data indexed through this driver so far, and waits for it to be handed off. Blocks until complete.
+   * Retries forever on transient failures, but may exit early on permanent failures.
+   *
+   * Should be called after all data has been added and persisted through {@link #add(InputRow, Supplier)} and
+   * {@link #persist(Committer)}.
+   *
+   * @param publisher publisher to use for this set of segments
+   * @param committer committer representing all data that has been added so far
+   *
+   * @return segments and metadata published if successful, or null if segments could not be handed off due to
+   * transaction failure with commit metadata.
+   */
+  public SegmentsAndMetadata finish(
+      final TransactionalSegmentPublisher publisher,
+      final Committer committer
+  ) throws InterruptedException
+  {
+    final SegmentsAndMetadata segmentsAndMetadata = publishAll(publisher, wrapCommitter(committer));
+
+    if (segmentsAndMetadata != null) {
+      log.info("Awaiting handoff...");
+      synchronized (handoffMonitor) {
```

Comment:
ok, added. the semantics are slightly different than in RealtimePlumber- here it only applies to historicals loading the data, not to push/publish. but I think this is actually more useful anyway, as the historicals are probably the thing we want to be timing out on, rather than push to deep storage.